Adds append, appendAll and prepend to Buffer

Prithvirajbilla opened the pull request:
Fixes https://github.com/scala/bug/issues/10970

reference to _append_ and _appendAll_ methods in Buffer (scala 2.12.x)

https://github.com/scala/scala/blob/2.12.x/src/library/scala/collection/mutable/BufferLike.scala#L145

Prithvirajbilla committed:
adds deprecated methods append and appendAll to Buffer

julienrf commented on src/library/scala/collection/mutable/Buffer.scala:
```diff
@@ -19,7 +19,19 @@ trait Buffer[A]
+    *  @param elems  the elements to append.
+    */
+  @deprecated("Use addAll instead", "2.13.0")
+  def append(elems: A*) { appendAll(elems) }
```
I suggest to not use the procedure syntax (which has been deprecated), and to forward to `addAll` directly:

~~~ scala
def append(elems: A*): Unit = addAll(elems)
def appendAll(xs: IterableOnce[A]): Unit = addAll(xs)
~~~

Prithvirajbilla replied:
updated. thanks.

szeiger commented on src/library/scala/collection/mutable/Buffer.scala:
```diff
@@ -19,7 +19,19 @@ trait Buffer[A]
     */
   def prepend(elem: A): this.type
 
-  @`inline` final def append(elem: A): this.type = addOne(elem)
```
Why remove this one? Users will end up with the deprecated (and less efficient) version instead. I don't think it makes sense to deprecate `append` when `prepend` is still supported.

julienrf replied:
Oh right, I forgot about that (see https://github.com/scala/scala/pull/6819 )…

So, do we want to have both `append` and `appendAll`, in addition to `addOne` and `addAll`, but for the sake of symmetry with `prepend` and `prependAll`?

szeiger commented on src/library/scala/collection/mutable/Buffer.scala:
```diff
@@ -19,7 +19,19 @@ trait Buffer[A]
+    *  @param elems  the elements to append.
+    */
+  @deprecated("Use addAll instead", "2.13.0")
+  def append(elems: A*): Unit = addAll(elems)
```
The deprecated methods should be `@inline final`. Otherwise they are prone to getting overridden (which may not have the intended semantics).

julienrf replied:
What about making them `@deprecatedOverriding` instead of `final` to not break existing overrides?

Prithvirajbilla replied:
Added `@deprecatedOverriding` and `@inline ` tag.

julienrf commented:
Hey @Prithvirajbilla, thanks a lot for your work. However, I gave you a wrong direction in the initial issue description and I’m sorry about that. I’ve explained in https://github.com/scala/bug/issues/10970#issuecomment-402637731 what I think we should do instead. We appreciate a lot if you still want to fix the issue, just let us know!

Prithvirajbilla commented:
I'd like to fix the issue. have gone through your comments. I will send a PR this weekend.

Prithvirajbilla commented:
Updated:

- [x] added `append` and `appendAll`, and made `addOne` and `addAll` final aliases to these methods. with same signature  `this.type`.
- [x] added deprecated variadic versions of `append` and `prepend` with same signature as `addOne` or `addAll` which is `this.type`.

julienrf commented on src/library/scala/collection/mutable/Buffer.scala:
```diff
@@ -19,13 +19,29 @@ trait Buffer[A]
+    *
+    *  @param elem  the element to append.
+    */
+  final def append(elem: A): this.type = addOne(elem)
```
You could keep the `@inline` annotation.

Prithvirajbilla replied:
updated

julienrf commented on src/library/scala/collection/mutable/Buffer.scala:
```diff
@@ -19,13 +19,29 @@ trait Buffer[A]
+  /** Appends the elements contained in a iterable object to this buffer.
+    *  @param xs  the iterable object containing the elements to append.
+    */
+  final def appendAll(xs: IterableOnce[A]): this.type = addAll(xs)
```
It should be `@inline` too.

Prithvirajbilla replied:
updated.

julienrf reviewed (commented):
That looks good! Can you make the aliases `@inline` and then squash the commits?

SethTisue commented:
this needs further work, see 
https://github.com/scala/bug/issues/11112

Prithvirajbilla commented:
what do you recommend that we do here? 

revert overloaded `append` method or fixing type inference issue.
